File: generated-code/codecraft/csharp/Model/PlayerView.cs
namespace TransGenTest.Model
{
    public struct PlayerView
    {
        public int MyId { get; set; }
        public int MapSize { get; set; }
        public bool FogOfWar { get; set; }
        public System.Collections.Generic.IDictionary<Model.EntityType, Model.EntityProperties> EntityProperties { get; set; }
        public int MaxTickCount { get; set; }
        public int MaxPathfindNodes { get; set; }
        public int CurrentTick { get; set; }
        public Model.Player[] Players { get; set; }
        public Model.Entity[] Entities { get; set; }
    
        public PlayerView(int myId, int mapSize, bool fogOfWar, System.Collections.Generic.IDictionary<Model.EntityType, Model.EntityProperties> entityProperties, int maxTickCount, int maxPathfindNodes, int currentTick, Model.Player[] players, Model.Entity[] entities)
        {
            this.MyId = myId;
            this.MapSize = mapSize;
            this.FogOfWar = fogOfWar;
            this.EntityProperties = entityProperties;
            this.MaxTickCount = maxTickCount;
            this.MaxPathfindNodes = maxPathfindNodes;
            this.CurrentTick = currentTick;
            this.Players = players;
            this.Entities = entities;
        }
    
        public static PlayerView ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new PlayerView();
            result.MyId = reader.ReadInt32();
            result.MapSize = reader.ReadInt32();
            result.FogOfWar = reader.ReadBoolean();
            int entityPropertiesSize = reader.ReadInt32();
            result.EntityProperties = new System.Collections.Generic.Dictionary<Model.EntityType, Model.EntityProperties>(entityPropertiesSize);
            for (int entityPropertiesIndex = 0; entityPropertiesIndex < entityPropertiesSize; entityPropertiesIndex++)
            {
                Model.EntityType entityPropertiesKey;
                Model.EntityProperties entityPropertiesValue;
                entityPropertiesKey = EntityTypeHelper.ReadFrom(reader);
                entityPropertiesValue = Model.EntityProperties.ReadFrom(reader);
                result.EntityProperties.Add(entityPropertiesKey, entityPropertiesValue);
            }
            result.MaxTickCount = reader.ReadInt32();
            result.MaxPathfindNodes = reader.ReadInt32();
            result.CurrentTick = reader.ReadInt32();
            result.Players = new Model.Player[reader.ReadInt32()];
            for (int playersIndex = 0; playersIndex < result.Players.Length; playersIndex++)
            {
                result.Players[playersIndex] = Model.Player.ReadFrom(reader);
            }
            result.Entities = new Model.Entity[reader.ReadInt32()];
            for (int entitiesIndex = 0; entitiesIndex < result.Entities.Length; entitiesIndex++)
            {
                result.Entities[entitiesIndex] = Model.Entity.ReadFrom(reader);
            }
            return result;
        }
    
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(MyId);
            writer.Write(MapSize);
            writer.Write(FogOfWar);
            writer.Write(EntityProperties.Count);
            foreach (var entityPropertiesEntry in EntityProperties)
            {
                var entityPropertiesKey = entityPropertiesEntry.Key;
                var entityPropertiesValue = entityPropertiesEntry.Value;
                writer.Write((int) (entityPropertiesKey));
                entityPropertiesValue.WriteTo(writer);
            }
            writer.Write(MaxTickCount);
            writer.Write(MaxPathfindNodes);
            writer.Write(CurrentTick);
            writer.Write(Players.Length);
            foreach (var playersElement in Players)
            {
                playersElement.WriteTo(writer);
            }
            writer.Write(Entities.Length);
            foreach (var entitiesElement in Entities)
            {
                entitiesElement.WriteTo(writer);
            }
        }
    
        public override string ToString() {
            string stringResult = "PlayerView { ";
            stringResult += "MyId: ";
            stringResult += MyId.ToString();
            stringResult += ", ";
            stringResult += "MapSize: ";
            stringResult += MapSize.ToString();
            stringResult += ", ";
            stringResult += "FogOfWar: ";
            stringResult += FogOfWar.ToString();
            stringResult += ", ";
            stringResult += "EntityProperties: ";
            stringResult += "{ ";
            int entityPropertiesIndex = 0;
            foreach (var entityPropertiesEntry in EntityProperties)
            {
                if (entityPropertiesIndex != 0) {
                    stringResult += ", ";
                }
                var entityPropertiesKey = entityPropertiesEntry.Key;
                stringResult += entityPropertiesKey.ToString();
                stringResult += ": ";
                var entityPropertiesValue = entityPropertiesEntry.Value;
                stringResult += entityPropertiesValue.ToString();
                entityPropertiesIndex++;
            }
            stringResult += " }";
            stringResult += ", ";
            stringResult += "MaxTickCount: ";
            stringResult += MaxTickCount.ToString();
            stringResult += ", ";
            stringResult += "MaxPathfindNodes: ";
            stringResult += MaxPathfindNodes.ToString();
            stringResult += ", ";
            stringResult += "CurrentTick: ";
            stringResult += CurrentTick.ToString();
            stringResult += ", ";
            stringResult += "Players: ";
            stringResult += "[ ";
            int playersIndex = 0;
            foreach (var playersElement in Players)
            {
                if (playersIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += playersElement.ToString();
                playersIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Entities: ";
            stringResult += "[ ";
            int entitiesIndex = 0;
            foreach (var entitiesElement in Entities)
            {
                if (entitiesIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += entitiesElement.ToString();
                entitiesIndex++;
            }
            stringResult += " ]";
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/codecraft/go/model/player_view.go
package model

import "io"
import . "trans_gen_test/stream"

type PlayerView struct {
    MyId int32
    MapSize int32
    FogOfWar bool
    EntityProperties map[EntityType]EntityProperties
    MaxTickCount int32
    MaxPathfindNodes int32
    CurrentTick int32
    Players []Player
    Entities []Entity
}

func NewPlayerView(myId int32, mapSize int32, fogOfWar bool, entityProperties map[EntityType]EntityProperties, maxTickCount int32, maxPathfindNodes int32, currentTick int32, players []Player, entities []Entity) PlayerView {
    return PlayerView {
        MyId: myId,
        MapSize: mapSize,
        FogOfWar: fogOfWar,
        EntityProperties: entityProperties,
        MaxTickCount: maxTickCount,
        MaxPathfindNodes: maxPathfindNodes,
        CurrentTick: currentTick,
        Players: players,
        Entities: entities,
    }
}

func ReadPlayerView(reader io.Reader) PlayerView {
    var myId int32
    myId = ReadInt32(reader)
    var mapSize int32
    mapSize = ReadInt32(reader)
    var fogOfWar bool
    fogOfWar = ReadBool(reader)
    var entityProperties map[EntityType]EntityProperties
    entityPropertiesSize := ReadInt32(reader)
    entityProperties = make(map[EntityType]EntityProperties)
    for entityPropertiesIndex := int32(0); entityPropertiesIndex < entityPropertiesSize; entityPropertiesIndex++ {
        var entityPropertiesKey EntityType
        entityPropertiesKey = ReadEntityType(reader)
        var entityPropertiesValue EntityProperties
        entityPropertiesValue = ReadEntityProperties(reader)
        entityProperties[entityPropertiesKey] = entityPropertiesValue
    }
    var maxTickCount int32
    maxTickCount = ReadInt32(reader)
    var maxPathfindNodes int32
    maxPathfindNodes = ReadInt32(reader)
    var currentTick int32
    currentTick = ReadInt32(reader)
    var players []Player
    players = make([]Player, ReadInt32(reader))
    for playersIndex := range players {
        var playersElement Player
        playersElement = ReadPlayer(reader)
        players[playersIndex] = playersElement
    }
    var entities []Entity
    entities = make([]Entity, ReadInt32(reader))
    for entitiesIndex := range entities {
        var entitiesElement Entity
        entitiesElement = ReadEntity(reader)
        entities[entitiesIndex] = entitiesElement
    }
    return PlayerView {
        MyId: myId,
        MapSize: mapSize,
        FogOfWar: fogOfWar,
        EntityProperties: entityProperties,
        MaxTickCount: maxTickCount,
        MaxPathfindNodes: maxPathfindNodes,
        CurrentTick: currentTick,
        Players: players,
        Entities: entities,
    }
}

func (playerView PlayerView) Write(writer io.Writer) {
    myId := playerView.MyId
    WriteInt32(writer, myId)
    mapSize := playerView.MapSize
    WriteInt32(writer, mapSize)
    fogOfWar := playerView.FogOfWar
    WriteBool(writer, fogOfWar)
    entityProperties := playerView.EntityProperties
    WriteInt32(writer, int32(len(entityProperties)))
    for entityPropertiesKey, entityPropertiesValue := range entityProperties {
        WriteInt32(writer, int32(entityPropertiesKey))
        entityPropertiesValue.Write(writer)
    }
    maxTickCount := playerView.MaxTickCount
    WriteInt32(writer, maxTickCount)
    maxPathfindNodes := playerView.MaxPathfindNodes
    WriteInt32(writer, maxPathfindNodes)
    currentTick := playerView.CurrentTick
    WriteInt32(writer, currentTick)
    players := playerView.Players
    WriteInt32(writer, int32(len(players)))
    for _, playersElement := range players {
        playersElement.Write(writer)
    }
    entities := playerView.Entities
    WriteInt32(writer, int32(len(entities)))
    for _, entitiesElement := range entities {
        entitiesElement.Write(writer)
    }
}